src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs

```diff
@@ -77,7 +93,23 @@ public static bool TryParseAsISO(ReadOnlySpan<char> source, out DateTimeOffset v
 
             int length = JsonReaderHelper.GetUtf8FromText(source, bytes);
 
-            return TryParseAsISO(bytes.Slice(0, length), out value);
+            bytes = bytes.Slice(0, length);
+
+            if (bytes.IndexOf(JsonConstants.BackSlash) != -1)
```

Comment:
Thinking about it, shouldn't these checks be pushed into the byte-based `TryParseAsISO` instead?

Reply:
I was thinking about that.  The byte-based methods are also called by `JsonDocument` and `Utf8JsonReader` where the source value has been already unescaped. So, I put these checks into the char-based `TryParseAsISO` to avoid double-checking.

But if you think it would be better to push the checks into the byte-based methods, I could create a separate PR.

Reply:
If you pushed it down into the byte-based overloads, would you be able to remove the code from the other call sites?

Reply:
I've looked at the code and realized that I wasn't entirely correct. The `TryGetEscapedDateTime(Offset)` methods also call the byte-based `TryParseAsISO`. 

If we move those checks into the byte-based methods and remove the code from the other call sites, we would have to change the `TryGetEscapedDateTime(Offset)` implementation. So they will be called only from `TryParseAsISO`. But in that case, we would always have to check for escaped characters, despite `JsonDocument` and `Utf8JsonReader` flags.

https://github.com/dotnet/runtime/blob/3d9a603e195ad583c7846265d97bd99e69a6f0d6/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs#L665-L669
https://github.com/dotnet/runtime/blob/3d9a603e195ad583c7846265d97bd99e69a6f0d6/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs#L849-L852

Reply:
> But in that case, we would always have to check for escaped characters, despite `JsonDocument` and `Utf8JsonReader` flags.

Right. I wouldn't change it if that requires redundant checks for escaping. If there is a cleaner refactor that can be done to avoid code duplication but avoid any extra/duplicate checks, then that would be great.